lightning/src/util/persist.rs

```diff
@@ -243,10 +250,889 @@ where
+					.map_err(|e| {
+						log_error!(
+							self.logger,
+							"Monitor update for monitor {} failed: {:?}",
+							monitor_name.as_str(),
+							e
+						);
+						io::Error::new(io::ErrorKind::Other, "Monitor update failed")
+					})?;
+			}
+			// ...push the result into the return vec
+			res.push((bh, monitor))
+		}
+		Ok(res)
+	}
+
+	/// Read a channel monitor.
+	fn read_monitor(
+		&self,
+		monitor_name: &MonitorName,
+	) -> io::Result<(
+		BlockHash,
+		ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+	)> {
+		let key = monitor_name.as_str();
+		let outpoint: OutPoint = monitor_name.try_into()?;
+		let mut monitor_cursor = io::Cursor::new(self.kv_store.read(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			key,
+		)?);
+		// Discard the sentinel bytes if found.
+		if monitor_cursor
+			.get_ref()
+			.starts_with(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL)
+		{
+			monitor_cursor.set_position(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() as u64);
+		}
+		match <(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>::read(
+			&mut monitor_cursor,
+			(&*self.entropy_source, &*self.signer_provider),
+		) {
+			Ok((blockhash, channel_monitor)) => {
+				if channel_monitor.get_funding_txo().0.txid != outpoint.txid
+					|| channel_monitor.get_funding_txo().0.index != outpoint.index
+				{
+					log_error!(
+						self.logger,
+						"ChannelMonitor {} was stored under the wrong key!",
+						monitor_name.as_str()
+					);
+					return Err(io::Error::new(
+						io::ErrorKind::InvalidData,
+						"ChannelMonitor was stored under the wrong key",
+					));
+				}
+				Ok((blockhash, channel_monitor))
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to read ChannelMonitor {}, reason: {}",
+					monitor_name.as_str(),
+					e,
+				);
+				Err(io::Error::new(
+					io::ErrorKind::InvalidData,
+					"Failed to read ChannelMonitor",
+				))
+			}
+		}
+	}
+
+	/// Read a channel monitor update.
+	fn read_monitor_update(
+		&self,
+		monitor_name: &MonitorName,
+		update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate> {
+		let sub_ns = monitor_name.as_str();
+		let key = update_name.as_str();
+		let update_bytes =
+			self.kv_store
+				.read(CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE, sub_ns, key)?;
+		ChannelMonitorUpdate::read(&mut io::Cursor::new(update_bytes)).map_err(|e| {
+			log_error!(
+				self.logger,
+				"Failed to read ChannelMonitorUpdate {}/{}, reason: {}",
+				monitor_name.as_str(),
+				update_name.as_str(),
+				e,
+			);
+			io::Error::new(
+				io::ErrorKind::InvalidData,
+				"Failed to read ChannelMonitorUpdate",
+			)
+		})
+	}
+
+	/// Lazily delete updates in the given range for the given monitor.
+	fn lazy_delete_updates(
+		&self,
+		monitor_name: &MonitorName,
+		first_update_id: u64,
+		last_update_id: u64,
+	) -> io::Result<()> {
+		for update_id in first_update_id..=last_update_id {
```

Comment:
Should we do a `list` here? Otherwise, the deep cleaning advice would be expensive if traversing the entire `u64` range.

Reply:
Earlier versions of this did a lot more listing (not just here, but at init time and elsewhere), but there was a lot of pushback (I think rightfully) on avoiding that since we are dealing with a monotonically increasing sequence and can compute the ranges by reading update_ids from the monitors. The RFC discussed this in a few threads, such as [this one](https://github.com/orgs/lightningdevkit/discussions/2545#discussioncomment-6971170).

This function is called from two places (part of the reason I didn't inline it... I'm also getting a lot of pushback about inlining, and I'd like to pull out more parts to unit test but 🤷 ). I don't think those places will allow the entire u64 range to be called here (the range computation is an example of something I want to unit test). That would be a severe error worth testing and avoiding, as that finite range is long enough that it'll take huge amounts of wall time and I/O to do.

Reply:
Ah, actually I'm mistaken. Passing `CLOSED_CHANNEL_UPDATE_ID` to `cleanup_stale_updates` will result in a range of `0..=monitor.get_latest_update_id()`. So not the whole u64 range, but still an increasing number of reads/deletes as the monitor is updated.

Can't we refactor this such that the loop body can be called both here and directly in `cleanup_stale_updates`? Seems like the latter doesn't really need a `search_depth`. It can `list` and delete anything that it no longer needs.

Reply:
Actually I think you are right, because `0..=monitor.get_latest_update_id()` will be the whole `u64` range if the monitor is for a closed channel. That's why I'm wondering if for an out-of-band cleanup, we should just bite the bullet and list.

Reply:
Right, I think agree with all these points. I think we should list on deep/out-of-band cleans, and some refactoring is in order here.